webapp/tests/karma/ts/services/create-user-for-contacts.service.spec.ts

```diff
@@ -0,0 +1,373 @@
+
+      expect(() => service.setReplaced(originalContact, NEW_CONTACT))
+        .to.throw('The new contact must have the same parent as the original contact when replacing a user.');
+    });
+
+    it('throws an error when the new contact does not have a parent', () => {
+      const newContact = Object.assign({}, NEW_CONTACT, { parent: {} });
+
+      expect(() => service.setReplaced(ORIGINAL_CONTACT, newContact))
+        .to.throw('The new contact must have the same parent as the original contact when replacing a user.');
+    });
+  });
+
+  describe('isReplaced', () => {
+    it('returns true when the given contact is replaced', () => {
+      const pendingContact = getContactWithStatus('PENDING');
+      expect(service.isReplaced(pendingContact)).to.be.true;
+    });
+
+    it('returns false when the given contact is not replaced', () => {
+      expect(service.isReplaced(ORIGINAL_CONTACT)).to.be.false;
+    });
+
+    it('returns false when the given contact has user_for_contact data but is not replaced', () => {
+      const originalContact = Object.assign({}, ORIGINAL_CONTACT, { user_for_contact: {} });
+      expect(service.isReplaced(originalContact)).to.be.false;
+    });
+  });
+
+  describe('getReplacedBy', () => {
+    it('returns the new contact when the given contact is replaced', () => {
+      const pendingContact = getContactWithStatus('PENDING');
+      expect(service.getReplacedBy(pendingContact)).to.equal(NEW_CONTACT._id);
+    });
+
+    it('returns undefined when the given contact is not replaced', () => {
+      expect(service.getReplacedBy(ORIGINAL_CONTACT)).to.equal(undefined);
+    });
+
+    it('returns false when the given contact has user_for_contact data but is not replaced', () => {
+      const originalContact = Object.assign({}, ORIGINAL_CONTACT, { user_for_contact: {} });
+      expect(service.isReplaced(originalContact)).to.be.false;
+    });
+  });
+
+  describe('syncStatusChanged', () => {
+    const SYNC_STATUS = { to: SyncStatus.Success, from: SyncStatus.Success };
+
+    it('updates contact with PENDING status to READY and logs out the user', async() => {
+      const pendingContact = getContactWithStatus('PENDING');
+      userSettingsService.get.resolves({ contact_id: pendingContact._id });
+      medicDb.get.withArgs(pendingContact._id).resolves(pendingContact);
+
+
+      expect(dbSyncService.subscribe.callCount).to.equal(1);
+      const syncStatusChanged = dbSyncService.subscribe.args[0][0];
+      await syncStatusChanged(SYNC_STATUS);
+
+      expect(userSettingsService.get.callCount).to.equal(1);
+      expect(dbService.get.callCount).to.equal(2);
+      expect(medicDb.get.callCount).to.equal(1);
+      expect(medicDb.get.args[0]).to.deep.equal([pendingContact._id]);
+      expect(medicDb.put.callCount).to.equal(1);
+      const updatedContact = Object.assign({}, pendingContact, {
+        user_for_contact: {
+          replaced: {
+            by: NEW_CONTACT._id,
+            status: 'READY'
```

Comment:
If only the status is changed, can this be reflected by using a deep copy and just editing that property? It would be more obvious to the reader.

Reply:
Looking at this close, I don't think this needs to be set anyway (since the `pendingContact` is actually being mutated anyway).  I have updated this test to actually just properly assert the `pendingContact.replaced` data.